frame/staking-rewards/src/lib.rs

```diff
@@ -0,0 +1,501 @@
+
+			let nft = T::get_protocol_nft::<StakingNFTOf<T>>(instance_id)?;
+
+			let now = T::Time::now();
+
+			let config = Self::get_config(&nft.asset)?;
+
+			// Possibly penalize the staked asset, protocol asset in this case.
+			let (penalized_stake, penalty_amount) =
+				nft.penalize_early_unstake_amount(now.as_secs(), config.early_unstake_penalty)?;
+
+			// Transfer back the (possibly penalized) staked amount.
+			let pallet_account = Self::account_id(&nft.asset);
+			// NOTE(hussein-aitlahcen): no need to keep pallet account alive.
+			T::Assets::transfer(nft.asset, &pallet_account, &to, penalized_stake, false)?;
+
+			// Move penalty to configured beneficiary
+			// NOTE(hussein-aitlahcen): no need to keep pallet account alive.
+			T::Assets::transfer(
+				nft.asset,
+				&pallet_account,
+				&config.penalty_beneficiary,
+				penalty_amount,
+				false,
+			)?;
+
+			// Decrement total shares.
+			TotalShares::<T>::try_mutate(nft.asset, |x| -> DispatchResult {
+				*x = x.safe_sub(&nft.stake.saturated_into::<u128>())?;
+				Ok(())
+			})?;
+
+			// Actually burn the NFT from the storage.
+			T::burn_protocol_nft::<StakingNFTOf<T>>(instance_id)?;
+
+			// Trigger event
+			Self::deposit_event(Event::<T>::Unstaked {
+				to: to.clone(),
+				stake: nft.stake,
+				penalty: penalty_amount,
+				nft: *instance_id,
+			});
+
+			Ok(())
+		}
+
+		fn claim(instance_id: &Self::InstanceId, to: &Self::AccountId) -> DispatchResult {
+			T::try_mutate_protocol_nft(instance_id, |nft: &mut StakingNFTOf<T>| {
+				let config = Self::get_config(&nft.asset)?;
+
+				let share = nft.stake.saturated_into::<u128>();
+				let total_shares = TotalShares::<T>::get(nft.asset);
+
+				// TODO(hussein-aitlahcen): extract pure maths to their own functions
+				let compute_reward =
+					|delta_collected: CollectedReward| -> Result<BalanceOf<T>, DispatchError> {
+						// Always increment but delta can't be > max supply <= u128.
+						let normalized_delta = u128::try_from(delta_collected)?;
+						nft.reward_multiplier
+							.mul_floor(normalized_delta)
+							.safe_mul(&share)?
+							.safe_div(&total_shares)?
+							.try_into()
+							.map_err(|_| ArithmeticError::Overflow.into())
+					};
+
+				let rewards = nft
+					.collected_rewards
+					.iter()
+					.map(|(reward_asset, previously_collected)| -> Result<(AssetIdOf<T>, BalanceOf<T>), DispatchError> {
+						match CollectedRewards::<T>::get(nft.asset, &reward_asset) {
+							Some(current_collected) => {
+								let delta_collected = current_collected.saturating_sub(*previously_collected);
+								let reward = compute_reward(delta_collected)?;
+								Ok((*reward_asset, reward))
+							},
+							None => Ok((*reward_asset, Zero::zero())),
+						}
+					})
+					.collect::<Result<Vec<_>, _>>()?;
+
+				let pallet_account = Self::account_id(&nft.asset);
+				for (reward_asset, reward) in rewards {
+					T::Assets::transfer(reward_asset, &pallet_account, to, reward, false)?;
```

Comment:
How does the protocol ensure that it has enough liquidity in all reward asset types to cover it's reward obligations? Or is that supposed to be managed though external means?

Reply:
i think it would be documented. there is no obligation. rewards are passed by any other protocol (or even person), if there are zero rewars - zero will be given (e.g. protocol did not earn nothing). but yeah, may be check for zero before transfer call?

Reply:
#picasso-chaos describes full protocol afaik

Reply:
> How does the protocol ensure that it has enough liquidity in all reward asset types to cover it's reward obligations? Or is that supposed to be managed though external means?

There is no obligation to reward the user, and the user has to manually claim the reward (meaning that he will know what's the available reward he will be able to claim)

Reply:
> There is no obligation to reward the user

I see, my question was more about say,
A user sees that they could claim some rewards based on the `nft.reward_multiplier` and calls `claim`, but the `pallet_account` does not have the necessary balance. So in that case user has to call again sometime later? Would this also mean that `unstake` calls would fail when there is no balance to pay the claimable reward?

Reply:
I'm not sure I follow, the reward multiplier is not meant to say whether you have some reward to claim or not?

Reply:
Based on the slack conversation, it's clear to me the rewards are funded by the protocols calling `StakingReward.transferReward`. Thanks!!